### PR TITLE
Add report generation for unit tests

### PR DESCRIPTION
**Pull request by talvbansal**
Report generation will show how well covered our codebase is. The reports however do not need to be commited to VC.

### CODE REVIEW

**Comment by talvbansal** [comment]
Reports will get generated into a folder called "reports/" which are already included within the .gitignore file, I've added phpunit being automatically run within the projects gulp file so that the checks and reports can be run automatically when using `gulp watch --production`.

**Comment by austintoddj** [comment]
@talvbansal Should we add the default `phpunit.xml` file to this PR? And just to be clear, how do either run the reports or not run the reports based on whatever the user wants?

**Comment by talvbansal** [comment]
@austintoddj, think of the `phpunit.xml.dist` a bit like a ready to run `.env.example` file. Its pre populated with a test suite config - in this instance it performs all the unit tests and generates a code coverage report. 

If you as the individual developer decided you didn't like its default behavior and say you wanted report generation to occur else where or you  wanted unit tests to stop on the first failed test you could copy the `phpunit.xml.dist` file to `phpunit.xml` and customise your test suite config without committing that to source control and changing behaviour for everyone else who may clone the project and run unit tests.

PHPUnit will look for a `phpunit.xml` first and then always fall back to `phpunit.xml.dist`.

So basically you need to decide if report generation for code coverage out of the box is something that you want and if so then accept this PR and if not then this is one you can reject. However adding `mix.phpUnit()` to the gulpfile is something I find incredibly useful when writing code / testing.

**Comment by austintoddj** [comment]
@talvbansal Ok pulled it down and checked it out. I liked it! Nice little printout of reports. So my only question now is, do you want/need to add the default `phpunit.xml` to be the primary for users on this PR? Then perhaps in the readme there needs to be a Unit Testing section telling developers why there are two of these files, and whichever they choose is up to them?

**Comment by talvbansal** [comment]
@austintoddj no you wouldn't ship a `phpunit.xml` file since there is no way to way to actually switch between the two configs. If `phpunit.xml` exists it will be run. If it doesn't exist `phpunit.xml.dist` will be run.

So a developer can choose to either build and run a customised test suite by choosing to creating thier own `phpunit.xml` which wont impact the core testing suite or leaving the project as it ships which will automatically run the `phpunit.xml.dist` file.

**Comment by austintoddj** [comment]
Ok, thanks for all the explanation, I think I'm finally grasping it. So lastly, do you get this error when you run the tests?

```
Runtime:       PHP 7.0.11
Configuration: /Users/taustin/Sites/canvas/phpunit.xml.dist
Error:         No code coverage driver is available
```

**Comment by talvbansal** [comment]
Aha you need the PHP-xdebug extension installed to generate code coverage! I completely forgot about that!

**Comment by austintoddj** [comment]
Wow, yeah that report is WAY more extensive than the little checkbox report that was being generated for me haha

**Comment by talvbansal** [comment]
Yeah i find it super useful to find code that hasn't been tested and code that isn't used anymore.

Hopefully you can see the value in the PR now :)

**Comment by austintoddj** [comment]
@talvbansal Yes for sure, lots of value there. And to answer my own question a while ago, if a dev doesn't want to take the time to let the reports generate during their own testing, all they'd need to do is comment the `<logging>` directive out of their `phpunit.xml.dist` file and they're back to normal. I'll merge this in right away!

**Comment by Naoray** [comment]
Just a quick question: is it possible that we start a documentation for canvas? I just discovered the new `reports` folder and didn't know what to do with it until I discovered this PR... It's pretty awesome to have such an overview about the code coverage. Is it possible we could somehow include this in a GUI?

**Comment by talvbansal** [comment]
@Naoray hey glad you like the reports. I'm not sure exposing the report data to the front end is going to be beneficial to end users although adding a note about what the reports folder is may be beneficial. 

We could always move the location of where the code coverage reports are generated to within the `tests` folder so maybe all of the test stuff is bundled together?

**Comment by Naoray** [comment]
@talvbansal guess this would be better.

**Comment by austintoddj** [comment]
@talvbansal @Naoray 

Like said above, I don't think exposing the reports to the frontend user would be too useful, especially because many users will never even run this, they will just use the blog for what it is.

I actually agree, let's move the `reports` directory inside `tests`. Nice to keep them together that way. 

As far as documentation, I have thought this over for some time and still haven't determined the best way to go about it. I could enable the Wiki feature on the repo, and allow others to help write some of it. I'd really love to use [Readme.io](http://readme.io), but it's a paid service...  Thoughts?

**Comment by Naoray** [comment]
@austintoddj 

A 'normal' user might never encounter the need of the reporting feature, but as a developer it is definitely  a nice to have feature. It could be 'hidden' from the other users somewhere in a `developer`'s setting.

Guess for the start github wiki would be sufficient.

**Comment by talvbansal** [comment]
@austintoddj I'll send a PR moving the reports inside of the tests folder now. 

I also agree with @Naoray to get things going we could start with the github wiki. Unless @austintoddj you could contact readme.io and ask about them sponsoring the project?

**Comment by austintoddj** [comment]
Thanks @talvbansal. I'll shoot them a message, can't hurt to try right?

**Comment by talvbansal** [comment]
@austintoddj you seem to have got a lot of sponsors for the project as it so so definitely try! Fingers crossed!

**Comment by austintoddj** [comment]
Message sent! 👍
